### PR TITLE
[WIP] Bring YARPC back

### PR DESCRIPTION
**Pull request by alsamylkin**
This PR is how YARPC module could look like with the DI refactor.
Caveat: we do need the new YARPC config, otherwise we'll need to implement converters
from map[string]interface{} to YARPC types ourselves.

### CODE REVIEW

**Comment by alsamylkin** [comment]
I know, that it is maybe a throwaway PR, but it does look like a very nice and simple module.

File: glide.yaml [hunk 1]
```diff
@@ -7,7 +7,7 @@ import:
 - package: github.com/gorilla/context
   version: ^1.1.0
 - package: go.uber.org/yarpc
-  version: ^v1.4.0
+  version: config
```
Comment (glibsm) [inline glide.yaml]: is this PR considered blocked until that branch merges?

File: modules/yarpc/yarpc.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+// Module represents a collection of YARPC components.
+type Module struct {
+	l           *zap.Logger
+	handlerCtor fx.Component
+	d           *yarpc.Dispatcher
+}
+
+// Transports represent a collection Procedures that will be registered with a dispatcher.
+type Transports struct {
```
Comment (glibsm) [inline modules/yarpc/yarpc.go]: what about middleware?

File: modules/yarpc/yarpc_test.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+		return &Transports{Ts: proc}, nil
+	}
+
+	module := New(fn)
+	for _, component := range module.Constructor() {
+		di.MustRegister(component)
+	}
+
+	cfg := map[string]interface{}{
```
Comment (glibsm) [inline modules/yarpc/yarpc_test.go]: i find your preference for this static syntax vs yaml baffling

File: modules/yarpc/yarpc_test.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+		},
+	}
+
+	provider := config.NewStaticProvider(cfg)
+	di.MustRegister(&provider)
+	tracer := opentracing.Tracer(opentracing.NoopTracer{})
+	di.MustRegister(&tracer)
+	di.MustRegister(&metrics.NopScope)
+	logger := zap.NewNop()
+	di.MustRegister(logger)
```
Comment (glibsm) [inline modules/yarpc/yarpc_test.go]: I thought our general guidance was "dig does not appear in tests".
You're still able to test the functions if you memorized them on a module, then you can call them directly after obtaining test/mock objects of tracer, logger, etc

File: glide.yaml (hunk 1 above)
Comment (alsamylkin) [inline glide.yaml]: yes, unless we want out own config parsing, which we should avoid.

File: modules/yarpc/yarpc.go (hunk 2 above)
Comment (alsamylkin) [inline modules/yarpc/yarpc.go]: We can have a constructor for it, not in this PR, but I don't see any problem adding it.
Middleware could also be populated from the yarpc's `Configurator.LoadConfig`.

File: modules/yarpc/yarpc_test.go (hunk 3 above)
Comment (alsamylkin) [inline modules/yarpc/yarpc_test.go]: I do like types and it is hard to see them in yaml. It also breaks my indentation if I am going to use ``.

File: modules/yarpc/yarpc_test.go (hunk 4 above)
Comment (alsamylkin) [inline modules/yarpc/yarpc_test.go]: Well, it depends what tests are we talking. I think I can replace dig here with a service, but it seems for module tests(not a user level functionality) we'll need dig sometimes.

Comment (glibsm) [inline modules/yarpc/yarpc_test.go]: Right, so the service tests definitely need to use dig. However, the modules might not necessarily need to.

Imagine that all the constructor functions were either named or memorized. Then, you can test them individually just by calling them direclty, without the need to involve dig, or a service.

**Comment by glibsm** [comment]
The yarpc module will make it's way to a new repo

**Pull request closed without merging**